src/base/utils/gzip.cpp

```diff
@@ -31,127 +31,190 @@
 
 #include <vector>
 
+#include <QtGlobal>
+#include <QBuffer>
 #include <QByteArray>
+#include <QIODevice>
 
 #ifndef ZLIB_CONST
 #define ZLIB_CONST  // make z_stream.next_in const
 #endif
 #include <zlib.h>
 
-QByteArray Utils::Gzip::compress(const QByteArray &data, const int level, bool *ok)
+bool Utils::Gzip::compress(QIODevice &source, QIODevice &dest, int level)
```

Comment:
FYI, boost::iostream provides interface for gzip compression/decompression. It seems to be a good idea to utilize it (directly at where you need it) instead of creating our own stream interface in utils/gzip.cpp
https://techoverflow.net/2020/01/13/how-to-gzip-compress-on-the-fly-in-c-using-boostiostreams/
https://www.boost.org/doc/libs/1_75_0/libs/iostreams/doc/classes/gzip.html#examples

Reply:
>This will introduce an additional dependency on boost-iostream. 

qbt already have dependency on boost.

Anyway, don't forget to modify/add test cases to [testutilsgzip.cpp](https://github.com/qbittorrent/qBittorrent/blob/master/test/testutilsgzip.cpp).

Reply:
> I think this issue can be discussed later.

:+1:

Reply:
> Anyway, don't forget to modify/add test cases to [testutilsgzip.cpp](https://github.com/qbittorrent/qBittorrent/blob/master/test/testutilsgzip.cpp).

Aren't the tests that already exist enough? They will use the new implementation indirectly anyway.

Reply:
>Aren't the tests that already exist enough? They will use the new implementation indirectly anyway.

When testing, you should treat the functions as black box and shouldn't assume its implementation. This PR introduce new public function so I expect test cases for it.

Reply:
I thought boost.iostreams was header-only and it isn't the case. So forget that idea.
https://www.boost.org/doc/libs/1_82_0/more/getting_started/windows.html#header-only-libraries